### PR TITLE
Implement Events Handling (#1)

### PR DESCRIPTION
**Pull request by ilslv**
Part of #1

## Synopsis

Firstly we should implement abstractions for `Event`s 

## Solution

- [x] Traits
  - [x] `Event`
  - [x] `VersionedEvent`
  - [x]  `EventSourced`
  - [x] `EventInitialised`
- [x] Structs
  - [x]  `InitialEvent` wrapper
- [x] Proc macros
  - [x] `Event` derive
  - [x] `VersionedEvent` derive
- [x] Toolchain
  - [x] `Makefile`
  - [x] PR and Issue templates
  - [x] GitHub Actions CI

## Checklist

- Created PR:
    - [x] In [draft mode][l:1]
    - [x] Name contains `Draft: ` prefix
    - [x] Name contains issue reference
    - [x] Has `k::` labels applied
    - [x] Has assignee
- [x] Documentation is updated (if required)
- [x] Tests are updated (if required)
- [x] Changes conform code style
- [x] CHANGELOG entry is added (if required)
- [x] FCM (final commit message) is posted
    - [x] and approved
- [x] [Review][l:2] is completed and changes are approved
- Before merge:
    - [x] Milestone is set
    - [x] PR's name and description are correct and up-to-date
    - [x] `Draft: ` prefix is removed
    - [x] All temporary labels are removed

[l:1]: https://help.github.com/en/articles/about-pull-requests#draft-pull-requests
[l:2]: https://help.github.com/en/articles/reviewing-changes-in-pull-requests

### CODE REVIEW

**Comment by ilslv** [comment]
## FCM

```
Implement Events Handling (#1, #2)

- Traits
  - Event
  - VersionedEvent
  - EventSourced
  - EventInitialised
- Structs
  - InitialEvent wrapper
  - EventVersion
- Proc macros
  - Event derive
  - VersionedEvent derive
- Toolchain
  - Makefile
  - GitHub Actions CI
```

File: core/src/event/mod.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+/// A sequence of [`Event`]s may represent a concrete versioned state of an
+/// Aggregate.
+///
+/// [Event Sourcing]: https://martinfowler.com/eaaDev/EventSourcing.html
+pub trait Event {
+    /// Returns type of this [`Event`].
+    ///
+    /// _Note:_ This should effectively be a constant value, and should never
+    /// change.
+    fn event_type(&self) -> &'static str;
```
Comment (tyranron) [inline core/src/event/mod.rs]: Let's change it to `name`. Termin `type` here was borrowed from other implementations, but tends to make more confusion. With `name` we have a clear separation when we're saying "event name" and "event type".

File: core/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    missing_debug_implementations,
+    missing_docs,
+    unreachable_pub,
+    unused_import_braces,
+    unused_labels,
+    unused_qualifications,
+    unused_results
+)]
+
+mod event;
```
Comment (tyranron) [inline core/src/lib.rs]: Let's hide it behind `es` (event sourcing) feature gate.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      - run: make cargo.fmt check=yes
+
+
+
+
+  ###########
+  # Testing #
+  ###########
+
+  test:
```
Comment (tyranron) [inline .github/workflows/ci.yml:56]: Also let's add tests for MSRV (minimum supported Rust version) from the start.

File: core/src/event/mod.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+/// Before items can be [`Sourced`], they need to be [`Initialized`].
+pub trait Initialized<Ev: ?Sized> {
+    /// Creates initial state from given [`Event`].
+    fn init(event: &Ev) -> Self;
+}
+
+/// Wrapper-type intended for [`Event`]s that can initialize [`Sourced`] items.
+#[derive(Debug, RefCast)]
+#[repr(transparent)]
+pub struct Initial<Ev: ?Sized>(pub Ev);
```
Comment (tyranron) [inline core/src/event/mod.rs]: `Clone`, `Copy` and transparent `Display` would be useful as well.

File: core/src/event/mod.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+impl<Ev: Event + ?Sized, Agg: Initialized<Ev>> Sourced<Initial<Ev>>
```
Comment (tyranron) [inline core/src/event/mod.rs]: It would be better to use `S` and "state" terminology here than "aggregate" as more general.

File: core/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+#[doc(inline)]
+pub use event::{
+    Event, Initial as InitialEvent, Initialized as EventInitialized,
+    Sourced as EventSourced, Versioned as VersionedEvent,
```
Comment (tyranron) [inline core/src/lib.rs]: And let's call the module `es` (event sourcing).

File: codegen/impl/src/event/versioned.rs [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+use std::convert::TryFrom;
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{spanned::Spanned as _, Result};
+use synthez::{ParseAttrs, ToTokens};
+
+use super::MAX_UNIQUE_EVENTS;
+
+/// Derives `arcana::VersionedEvent` for struct.
```
Comment (tyranron) [inline codegen/impl/src/event/versioned.rs]: Use links to `arcana-core` crate. The crates layout allows it.

File: codegen/impl/src/event/versioned.rs [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    Ok(quote! { #definitions })
+}
+
+#[derive(ToTokens)]
+#[to_tokens(append(impl_from, unique_event_type_and_ver))]
+struct Definitions {
+    ident: syn::Ident,
+    generics: syn::Generics,
+    event_type: syn::LitStr,
+    event_ver: syn::LitInt,
```
Comment (tyranron) [inline codegen/impl/src/event/versioned.rs]: These fields should have documantation.

File: codegen/impl/src/event/versioned.rs [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    }
+}
+
+#[derive(Default, ParseAttrs)]
+struct Attrs {
+    #[parse(value)]
+    r#type: Option<syn::LitStr>,
+
+    #[parse(value)]
+    version: Option<syn::LitInt>,
```
Comment (tyranron) [inline codegen/impl/src/event/versioned.rs]: It would be nice to see `Attrs` defined before `Definition`, following parse -> generate intuitive logic.

File: src/private.rs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+            );
+        };
+    }
+
+    #[doc(hidden)]
+    #[must_use]
+    pub const fn all_unique<const N: usize>(
+        types: [Option<(&str, u16)>; N],
+    ) -> bool {
+        const fn str_eq(l: &str, r: &str) -> bool {
```
Comment (tyranron) [inline src/private.rs]: It's OK for this function to be outside.

**Review by tyranron — CHANGES_REQUESTED** [review]
@ilslv will discuss stuff additionally in videocall.

File: codegen/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+[package]
+name = "arcana-codegen"
+version = "0.1.0"
```
Comment (tyranron) [inline codegen/Cargo.toml]: This should be `0.1.0-dev`.

File: src/private.rs [hunk 12]
```diff
@@ -127,4 +113,23 @@ pub mod unique_event_type_and_ver {
 
         true
     }
+
+    #[doc(hidden)]
+    #[must_use]
+    const fn str_eq(l: &str, r: &str) -> bool {
```
Comment (tyranron) [inline src/private.rs]: No `#[doc(hidden)]`. Describe it. And an appropriate `// TODO:` for future removal.

File: core/src/es/mod.rs [hunk 13]
```diff
@@ -80,8 +79,8 @@ impl Version {
 }
 
 impl<Ev: Versioned> Event for Ev {
```
Comment (tyranron) [inline core/src/es/mod.rs]: `?Sized` here.

File: src/private.rs [hunk 14]
```diff
@@ -17,55 +17,58 @@ pub use static_assertions as sa;
...
+/// [`Event::name()`]: trait@crate::Event::name()
+/// [`Event::ver()`]: trait@crate::Event::ver()
 /// [`VersionedEvent`]: trait@crate::VersionedEvent
-pub mod unique_event_type_and_ver {
+pub mod unique_event_name_and_ver {
     #[doc(hidden)]
     #[macro_export]
-    macro_rules! unique_event_type_and_ver_for_struct {
-        ($max_events:literal, $event_type:literal, $event_ver:literal) => {
+    macro_rules! unique_event_name_and_ver_for_struct {
```
Comment (tyranron) [inline src/private.rs]: This macros better move to proc-macro machinery. There is no point to generate a half of an implementation via declarative and another part via proc macros. Better to keep things in a single place asap.

**Review by tyranron — CHANGES_REQUESTED** [review]
@ilslv still, discussion is required.

File: codegen/core/src/event/versioned.rs [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+            ));
+        }
+
+        let attrs = Attrs::parse_attrs("event", &input)?;
+        let (event_name, event_ver) = match (attrs.name, attrs.version) {
+            (Some(event_name), Some(event_ver)) => (event_name, event_ver),
+            _ => {
+                return Err(syn::Error::new_spanned(
+                    input,
+                    "`name` and `version` arguments expected",
```
Comment (tyranron) [inline codegen/core/src/event/versioned.rs]: We have `synthez::Required` for that.

File: codegen/core/src/event/versioned.rs [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+            #[automatically_derived]
+            impl #impl_generics ::arcana::codegen::UniqueEvents for
+                #name #ty_generics #where_clause
+            {
+                const COUNT: usize = 1;
+            }
+
+            impl #impl_generics #name #ty_generics #where_clause {
+                #[automatically_derived]
+                pub const fn __arcana_events() -> [(&'static str, u16); 1] {
```
Comment (tyranron) [inline codegen/core/src/event/versioned.rs]: It's better to hide such inner machinery from docs via `#[doc(hidden)]`.

**Review by tyranron — CHANGES_REQUESTED** [review]
Discussed.

File: codegen/core/src/es/event/versioned.rs [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+        let input = parse_quote! {
+            #[event(name = "event")]
+            struct Event;
+        };
+
+        let err = super::derive(input).unwrap_err();
+
+        assert_eq!(
+            format!("{}", err),
+            "either `version` or `ver` argument of `#[event]` attribute is \
```
Comment (ilslv) [inline codegen/core/src/es/event/versioned.rs]: @tyranron it looks like error message isn't deterministic. Sometimes I get `either 'ver' or 'version' argument of '#[event]' attribute is expected to be present, but is absent`, so `ver` and `version` are switched (even on 2 CI runs ([first](https://github.com/arcana-rs/arcana/actions/runs/1165386827), [second](https://github.com/arcana-rs/arcana/actions/runs/1165386785)) for the same code different jobs failed). As I understand this happens because `synthez` uses `HashSet<syn::Ident>` under the hood. `HashSet` docs are saying that there is requirement

> k1 == k2 -> hash(k1) == hash(k2)

And for `syn::Ident` that promise isn't held up as I understand. [`PartialEq` impl](https://docs.rs/proc-macro2/1.0.28/src/proc_macro2/lib.rs.html#949-953) compares ident itself and it's `Span`, while [`Hash` impl](https://docs.rs/proc-macro2/1.0.28/src/proc_macro2/lib.rs.html#978-982) simply hashes ident, ignoring it's `Span`. This can be easily fixed specifically for error message generation, but I'm not sure if this thing can cause bigger ploblems.

File: codegen/core/Cargo.toml [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+    "Ilya Solovyiov <ilya.solovyiov@gmail.com>",
+    "Kai Ren <tyranron@gmail.com>"
+]
+documentation = "https://docs.rs/arcana-codegen-core"
+homepage = "https://github.com/arcana-rs/arcana/tree/master/codegen/core"
+repository = "https://github.com/arcana-rs/arcana"
+readme = "README.md"
+
+[dependencies]
+arcana-core = { version = "0.1.0-dev", path = "../../core", features = ["es"] }
```
Comment (tyranron) [inline codegen/core/Cargo.toml]: We shouldn't bloat dependencies with unused ones. If it's required for docs, let's introduce an appropriate feature, disabled by default.

File: codegen/core/src/es/event/mod.rs [hunk 19]
```diff
@@ -0,0 +1,518 @@
...
+
+                    #event_array_population
+
+                    res
+                }
+            }
+
+            ::arcana::codegen::sa::const_assert!(
+                !::arcana::codegen::unique_events::has_duplicates(
+                    #name::__arcana_events()
```
Comment (tyranron) [inline codegen/core/src/es/event/mod.rs]: That simply won't work with generics.

**Review by tyranron — CHANGES_REQUESTED** [review]
I've accidentally started reviewing it without realizing it's still wip. Review and consider my changes.

File: codegen/core/src/es/event/mod.rs (hunk 19 above)
Comment (ilslv) [reply codegen/core/src/es/event/mod.rs]: @tyranron it looks like there is no way to make it work with generics for now 😔
With your solution compiler error is

> can't use generic parameters from outer function [E0401](https://doc.rust-lang.org/error-index.html#E0401)
> use of generic parameter from outer function

File: codegen/shim/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+/// enum AnyEvent {
+///     Chat(ChatEvent),
+///     File(FileEvent),
+/// }
+///
+/// // This fails to compile as contains `FileEvent` duplicated.
+/// #[derive(Event)]
+/// enum DuplicatedEvent {
+///     Any(AnyEvent),
+///     File(FileEvent),
```
Comment (tyranron) [inline codegen/shim/src/lib.rs:76]: @ilslv it seems that exactly this situation should be legit.

I guess it should be allowed to use the same event type multiple times. This allows us easily to merge sets of events.

What should be forbidden really - is using the same `event_name` + `event_ver` pair in __different__ Rust types.

Let's merge this PR as it is, and resolve the described issue in a separate one.

**Review by tyranron — APPROVED** [review]
## FCM

```
Provide basic events sourcing (#2, #1)

- bootstrap project structure

Additionally:
- bootstrap CI pipeline
- bootstrap Makefile
```
